pkg/ratelimiter/buckets.go

```diff
@@ -0,0 +1,80 @@
+package ratelimiter
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type Buckets struct {
+	log     *zap.Logger
+	mutex   sync.RWMutex
+	buckets map[string]*Entry
+}
+
+func NewBuckets(log *zap.Logger) *Buckets {
+	return &Buckets{
+		log:     log,
+		buckets: make(map[string]*Entry),
+		mutex:   sync.RWMutex{},
+	}
+}
+
+func (b *Buckets) getAndRefill(bucket string, limit *Limit, multiplier uint16, create bool) *Entry {
+	// The locking strategy is adapted from the following blog post: https://misfra.me/optimizing-concurrent-map-access-in-go/
+	b.mutex.RLock()
+	currentVal, exists := b.buckets[bucket]
+	b.mutex.RUnlock()
+	if !exists {
+		if !create {
+			return nil
+		}
+		b.mutex.Lock()
+		currentVal, exists = b.buckets[bucket]
```

Comment:
We didn't use to do the second check in the original code, but I think it needs to be there.